tools/data_source/upload.py

```diff
@@ -196,24 +197,24 @@ def add_composite_file(dataset, registry, output_path, files_path):
         datatype = registry.get_datatype_by_extension(dataset.file_type)
 
     def to_path(path_or_url):
-        is_url = path_or_url.find('://') != -1  # todo fixme
-        if is_url:
+        isa_url = is_url(path_or_url)
```

Comment:
I think we need to allow additional schemes here (or fall back to `://`), since `stream_url_to_file` is also called to realize `file://`  (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/sniff.py#L59) and gx*:// URIs (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/sniff.py#L56).

I think the most explicit approach here would be `if is_url(path_or_url, allow_list=['http://', 'https://', 'ftp://', 'file://']) or file_sources and file_sources.looks_like_uri(path)`.

Reply:
Ah,`file://` is sanitized both in fetch_data (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/api/_fetch_util.py#L36) and upload1 (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/parameters/grouping.py#L377), but the file source URIs still need to be allowed. In fact we may want to raise an `RequestParameterInvalidException` exception if `file://` is in `path_or_url`

Reply:
> but the file source URIs still need to be allowed. 

change should be easy .. but this one would need a chown :) 

> In fact we may want to raise an RequestParameterInvalidException exception if file:// is in path_or_url

where would you suggest to raise this? In `is_url`? Maybe we add a `deny_list` that leads to this exception.

Reply:
> but this one would need a chown :)

Shouldn't that be treated exactly like a path ?



> In `is_url`?

is_url should be generic, I'd raise it in stream_url_to_file (which will work with `file://` just fine).

Reply:
I'm confused. I guess it would be easier if you just commit? .. sorry..

Reply:
Done in f85086f6d5283d650b1ee56e33ccc1f6ad643a72. I was going to add a test, but it looks like we do have tests at least for `fetch_data` in https://github.com/bernt-matthias/galaxy/blob/f85086f6d5283d650b1ee56e33ccc1f6ad643a72/test/integration/test_remote_files.py#L98.

I'm not sure how we're doing uploads now (upload1 or data_fetch?) for file sources data in the UI, but that seems to be broken regardless. It just sends an empty POST request to `api/tools` ... so that's something to fix in another PR.